### DISCUSSION TITLE
How can I avoid full page reloads?

### DISCUSSION THREAD

**Discussion by devuxer** [vitejs/vite-plugin-react#112] [Q&A] [answered]
I seem to get a full page reload no matter what I do.

Add a console.log statement to a component? Full page reload.
Remove a console.log statement from a component? Full page reload.
Change a button label? Full page reload.
Breath on the keyboard? Full page reload. (okay, I kid)

Any idea what could cause this?

At the moment, I'm including the react plugin in my vite.config with zero arguments. Here is what I have for plugins:
```ts
plugins: [
    react(),
    viteTsconfigPaths(),
    svgrPlugin(),
]
```

In my console, I see messages like "hmr update" and "hmr invalidate". I only seem to get "hmr invalidate" when I have an error in my code (in which case, I'd expect a full page refresh), but even when it shows "hmr update", I *still* get a full page refresh. I thought HMR surgically re-renders only the component I edited (and its children), not the whole app.

**Comment by ArnaudBarre** [comment, marked as answer]
You should update to the latest version of the plugin you will get a message on why the file was invalidated. 
There are rules to follow in React to get HMR working: https://github.com/vitejs/vite-plugin-react/tree/main/packages/plugin-react#consistent-components-exports
If you're using eslint, you can try https://github.com/ArnaudBarre/eslint-plugin-react-refresh to find some issues

**Reply by devuxer** [reply]
@ArnaudBarre, Thanks for your response. I just started using VITE and this plugin, so I definitely have the latest (at least within a few days), however, the lint rule did turn up a warning for something that is probably the cause, but I'm not sure how to work around it.

I have a lazy import at the very top of my app, in index.tsx. Here is the code:

```ts
import { lazy, Suspense } from "react";
import { createRoot } from "react-dom/client";
import { loadAppData } from "./app-data";
import AboutPanel from "./features/about-panel/AboutPanel";

export const appData = loadAppData(); // sychronously load data
const rootElement = document.getElementById("root");
if (!rootElement) throw new Error(`HTML file must contain an element with ID "root"`);
const root = createRoot(rootElement);
const App = lazy(async () => import("./App"); 
root.render(
    <Suspense fallback={<AboutPanel isLoading />}>
        <App />
    </Suspense>,
);
``` 

I get a yellow squiggle under `App` with this warning:

> Fast refresh only works when a file only export components. Move your component(s) to a separate file.

I'm not sure what it means to put this in another file.  I did try putting the lazy call into a function instead of setting to a const, but I wasn't able to get this to work.

Any thoughts on how I could work around this issue? Basically, I want to load a little app data (which is pretty quick) then show a "loading screen" until the (very large) App component is rendered.

**Reply by devuxer** [reply]
Okay, I think I figured it out!

I added an `export` to `const App = ...`, which caused the yellow squiggle under `App` to disappear, but then a different yellow squiggle appeared for `export const appData...`, so, I moved that to `./app-data` and imported `appData` into my index.tsx file (instead of the function) so it would load before `<App />`.

Here's what I ended up with:

```ts
import { appData } from "@/app-data";
import { lazy, Suspense } from "react";
import { createRoot } from "react-dom/client";
import AboutPanel from "./features/about-panel/AboutPanel";

if (!appData) throw new Error("App data failed to load");
const rootElement = document.getElementById("root");
if (!rootElement) throw new Error(`HTML file must contain an element with ID "root"`);
const root = createRoot(rootElement);
export const App = lazy(async () => import("./App"));
root.render(
    <Suspense fallback={<AboutPanel isLoading />}>
        <App />
    </Suspense>,
);
```

Now I'm finally experiencing the full awesomeness of VITE. I make a change, and it's so instant, I question whether anything even happened (in a good way).

These rules will take a bit of getting used to, kind of like the "rules of hooks", but I really only had a handful of things in my code I had to fix, so no real harm to my DX.

Thanks again for pointing me in the right direction.

**Reply by ArnaudBarre** [reply]
That strange that you needed to re-export from your entrypoint.
The entrypoint that mount the react root should normally not contains export and be as small as possible because any change in it it trigger a full page reload.

Also making the full app lazy loose the benefice of lazy loading because you will the app chunk to do anything, expect if your app as a lot not written in React or multiple roots. If not I would advice not using lazy for the app and moving the lazy parts for secondary views

**Reply by devuxer** [reply]
I was using lazy because I wanted to show my about/loading panel using Suspense while the App loads (which takes about 5 seconds), not because I want the App component to load on demand (it will of course load every time).

I'm open to suggestions on how to do this a different way, though, if you think it's harming user or developer experience. Now that I've got Fast Refresh working, though, I'm not sure I see any downside.

**Reply by ArnaudBarre** [reply]
For that personally what I'm using it a small piece of html with the inline style in the root element that get replace when the app mount. So you get a working spinner in the first bytes of your page

**Reply by devuxer** [reply]
Everything is working perfectly now, though. I am no longer experiencing full-page reloads. If I make a change, it updates immediately in the browser.

A couple notes on my situation:

1. This is the front-end of an Electron desktop app, so bundle size, code splitting, etc. is not a priority.
2. My Fallback is a non-trivial React component, not just an icon/spinner, so it would not be straightforward to just put that in the HTML.

**Reply by ArnaudBarre** [reply]
Ok sounds good in that case 👍

**Reply by devuxer** [reply]
Thanks for your help!

**Comment by eisa-rezaei** [comment]
Hi,I have same issue, I didn't quite catch what should i do to prevent all page refresh could you provide me with an answer ? it takes lots of my time for any change.
